Source/TaurusTLS_Random.pas

```diff
@@ -0,0 +1,573 @@
+      {$IFDEF USE_INLINE}inline;{$ENDIF}
+
+    ///  <summary>
+    ///  The <c>PrivateRandom</c> <c>class property</c> reterns an instance of <see cref="TTaurusTLS_OSSLRandom" /> class
+    ///  which uses <c>openssl private</c> random bytes generator.
+    ///  </summary>
+    ///  <rematrks>
+    ///  User code <b>must not</b> free the obtained instance.
+    ///  </remarks>
+    class property PrivateRandom: TTaurusTLS_Random read FPrivateRandom;
+    ///  <summary>
+    ///  The <c>PrivateRandom</c> <c>class property</c> reterns an instance of <see cref="TTaurusTLS_OSSLRandom" /> class
+    ///  which uses <c>openssl public</c> random bytes generator.
+    ///  </summary>
+    ///  <rematrks>
+    ///  User code <b>must not</b> free the obtained instance.
+    ///  </remarks>
+    class property PublicRandom: TTaurusTLS_Random read FPublicRandom;
+  end;
+
+  ///  <summary>
+  ///  This exception indicates error returned from <c>openssl</c> random function(s)
+  ///  in the <see cref="TTaurusTLS_Random" /> class.
+  ///  </summary>
+  ERandom = class(ETaurusTLSAPICryptoError);
+
+implementation
+
+uses
+  TaurusTLSHeaders_err;
+
+{ TTaurusTLS_CustomOSSLRandomBytes }
+
+constructor TTaurusTLS_CustomOSSLRandomBytes.Create(ACtx: POSSL_LIB_CTX;
+  AStrength: TIdC_UINT);
+begin
+  FCtx:=ACtx;
+  FStrength:=AStrength;
+end;
+
+function TTaurusTLS_CustomOSSLRandomBytes.Random(var buf;
+  num: TIdC_SIZET): TIdC_INT;
+begin
+  if num > 0 then
+    Result:=DoRandom(FCtx, buf, num, FStrength)
+  else
+    Result:=1;
+end;
+
+{ TTaurusTLS_OSSLPrivateRandomBytes }
+
+function TTaurusTLS_OSSLPrivateRandomBytes.DoRandom(
+  ctx: POSSL_LIB_CTX; var buf; num: TIdC_SIZET; strength: TIdC_UINT): TIdC_INT;
+begin
+  Result:=RAND_priv_bytes_ex(ctx, @buf, num, strength);
+end;
+
+{ TTaurusTLS_OSSLPublicRandomBytes }
+
+function TTaurusTLS_OSSLPublicRandomBytes.DoRandom(
+  ctx: POSSL_LIB_CTX; var buf; num: TIdC_SIZET; strength: TIdC_UINT): TIdC_INT;
+begin
+  Result:=RAND_bytes_ex(ctx, @buf, num, strength);
+end;
+
+{ TTaurusTLS_OSSLRandom }
+
+class constructor TTaurusTLS_OSSLRandom.Create;
+begin
+  FPrivateRandom:=NewRandom(TTaurusTLS_OSSLPrivateRandomBytes.Create(nil, 0));
+  FPublicRandom:=NewRandom(TTaurusTLS_OSSLPublicRandomBytes.Create(nil, 0));
+end;
+
+class destructor TTaurusTLS_OSSLRandom.Destroy;
+begin
+  FreeAndNil(FPublicRandom);
+  FreeAndNil(FPrivateRandom);
+end;
+
+class function TTaurusTLS_OSSLRandom.NewRandom(
+  ARandomBytes: TTaurusTLS_CustomOSSLRandomBytes): TTaurusTLS_OSSLRandom;
+begin
+  Result:=Create(ARandomBytes);
+end;
+
+constructor TTaurusTLS_OSSLRandom.Create;
+begin
+  Assert(False, ClassName+' can not be creates with this constructor.');
+end;
+
+constructor TTaurusTLS_OSSLRandom.Create(ARandomGen: TTaurusTLS_CustomOSSLRandomBytes;
+  ACtx: POSSL_LIB_CTX; AStrength: TIdC_UINT);
+begin
+  Assert(Assigned(ARandomGen), 'ARandomGen must not be ''nil''.');
```

Comment:
Should create your own exception type derived from ETaurusTLSError. String constants should be in TaurusTLS_ResourceStrings.